gasstation/gasstattion.go

```diff
@@ -103,3 +115,116 @@ func (gs *GasStation) SuggestGasPrice() (uint64, error) {
+		height := lastBlock - i
+		if blkFee, ok := gs.feeCache.Get(height); ok {
+			// cache hit
+			log.T(ctx).Debug("fee cache hit", zap.Uint64("height", height))
+			bf := blkFee.(*blockFee)
+			baseFees[i] = bf.baseFee
+			gasUsedRatios[i] = bf.gasUsedRatio
+			blobBaseFees[i] = bf.blobBaseFee
+			blobGasUsedRatios[i] = bf.blobGasRatio
+		} else {
+			// read block fee from dao
+			log.T(ctx).Debug("fee cache miss", zap.Uint64("height", height))
+			blk, err := gs.dao.GetBlockByHeight(height)
+			if err != nil {
+				return 0, nil, nil, nil, nil, nil, status.Error(codes.NotFound, err.Error())
+			}
+			if i == 0 {
+				lastBlk = blk
+			}
+			baseFees[i] = blk.BaseFee()
+			gasUsedRatios[i] = float64(blk.GasUsed()) / float64(g.BlockGasLimitByHeight(blk.Height()))
+			blobBaseFees[i] = protocol.CalcBlobFee(blk.ExcessBlobGas())
+			blobGasUsedRatios[i] = float64(blk.BlobGasUsed()) / float64(params.MaxBlobGasPerBlock)
+			gs.feeCache.Add(height, &blockFee{
+				baseFee:      baseFees[i],
+				gasUsedRatio: gasUsedRatios[i],
+				blobBaseFee:  blobBaseFees[i],
+				blobGasRatio: blobGasUsedRatios[i],
+			})
+		}
+		// block priority fee percentiles
+		if len(rewardPercentiles) > 0 {
+			if blkPercents, ok := gs.percentileCache.Get(height); ok {
+				log.T(ctx).Debug("percentile cache hit", zap.Uint64("height", height))
+				rewards = append(rewards, feesPercentiles(blkPercents.(*blockPercents).ascEffectivePriorityFees, rewardPercentiles))
+			} else {
+				log.T(ctx).Debug("percentile cache miss", zap.Uint64("height", height))
+				receipts, err := gs.dao.GetReceipts(height)
+				if err != nil {
+					return 0, nil, nil, nil, nil, nil, status.Error(codes.NotFound, err.Error())
+				}
+				fees := make([]*big.Int, 0, len(receipts))
+				for _, r := range receipts {
+					fees = append(fees, r.PriorityFee())
+				}
+				sort.Slice(fees, func(i, j int) bool {
+					return fees[i].Cmp(fees[j]) < 0
+				})
+				rewards = append(rewards, feesPercentiles(fees, rewardPercentiles))
+				gs.percentileCache.Add(height, &blockPercents{
+					ascEffectivePriorityFees: rewards[i],
+				})
+			}
+		}
+	}
+	// fill next block base fee
+	if lastBlk == nil {
+		blk, err := gs.dao.GetBlockByHeight(lastBlock)
+		if err != nil {
+			return 0, nil, nil, nil, nil, nil, status.Error(codes.NotFound, err.Error())
+		}
+		lastBlk = blk
+	}
+	baseFees[blocks] = protocol.CalcBaseFee(g.Blockchain, &protocol.TipInfo{
+		Height:  lastBlock,
+		GasUsed: lastBlk.GasUsed(),
+		BaseFee: lastBlk.BaseFee(),
+	})
+	blobBaseFees[blocks] = protocol.CalcBlobFee(protocol.CalcExcessBlobGas(lastBlk.ExcessBlobGas(), lastBlk.BlobGasUsed()))
+	return lastBlock - blocks + 1, rewards, baseFees, gasUsedRatios, blobBaseFees, blobGasUsedRatios, nil
+}
+
+func feesPercentiles(ascFees []*big.Int, percentiles []float64) []*big.Int {
+	res := make([]*big.Int, len(percentiles))
+	for i, p := range percentiles {
+		idx := int(float64(len(ascFees)) * p)
+		if idx >= len(ascFees) {
+			idx = len(ascFees) - 1
```

Comment:
is it possible this `if` can enter multiple times in the loop?

Reply:
It should only happens when p equlas 1, b/c the convertion from float to int will truncate the decimal part in Golang